Make Galaxy (almost) packageable/installable

natefoo opened the pull request:
# 2019 Update

This won't be entirely ready for `pip install galaxy` but does what's needed to quickly finish it up. The plan is for the [galaxy-web-apps package](https://github.com/galaxyproject/galaxy/tree/dev/packages/web_apps/) package essentially a full install of Galaxy with its web stack, and the [galaxy metapackage](https://github.com/galaxyproject/galaxy/tree/dev/packages/meta/) will be pinned to stable releases. Because the packagification of Galaxy began just before updating this PR happened, the method originally used here (one big package) is no longer valid, and the metapackage isn't working yet. That will happen in a subsequent PR.

This PR includes some important useful pieces:
- `GalaxyConfiguration.config_dir` is `dirname(config_file)` (or `.` if there is no config file)
- `GalaxyConfiguration.data_dir` is `./data`
- No assumption that `cwd` is `galaxy_root`
- No assumption that configs are in `./config/`
- Sample configs are no longer required
- Galaxy now finds templates, static, and tools via symlinks in the galaxy lib dir (but these are not packageable yet)
- Sample configs moved to `lib/galaxy/config/sample`

The exception to the changes above: if running from `cwd = galaxy_root`, then `config_dir` and `data_dir` are `<cwd>/config` and `<cwd>/database` as before.

## Updated TODO list for the next PR
- [ ] Finish the metapackage
- [ ] Update `galaxy-config` for YAML configs
- [ ] Add `galaxy-uwsgi` or `galaxy -m uwsgi`
- [ ] Package `static/`
- [ ] Other directories: `external_services/`, `tool-data/`, ...
- [ ] Create a `galaxy-set-metadata` command to replace the tempfile method currently used
- [ ] Move `galaxy-manage-db` to galaxy-app after #8293 I think
- [ ] Fix tool migrations
- [ ] Other things I'm not thinking of

# Original PR

Building on #428, this attempts to make Galaxy installable via the standard Pythonic mechanism. To use:
1. Make a virtualenv and activate it
2. `PYTHONPATH= pip install --upgrade pip wheel`
3. `make wheels`
4. `PYTHONPATH= pip install -i https://wheels.galaxyproject.org/simple/ 'dist/galaxy-16.7.dev0-py2-none-any.whl[postgresql]'`
5. Create a directory somewhere, copy `config/galaxy.ini.sample` to it as `galaxy.ini`
6. `galaxy-paster serve galaxy.ini`

Alternatively:
1. Make a virtualenv and activate it
2. `PYTHONPATH= pip install --upgrade pip`
3. `make client`
4. `PYTHONPATH= pip install -i https://wheels.galaxyproject.org/simple '.[postgresql]'` (add `-e` for editable/develop mode)
5. As above
## TODO list
- [x] Figure out what to do with `lib/galaxy/datatypes/converters/`, `lib/galaxy/datatypes/set_meatadata_tool.xml`, etc. (just use `__file__` rather than assuming cwd is galaxy_root?)
- [x] Fix `galaxy-main`
- [x] Make a `galaxy-config` to create a Galaxy config directory and relevant contents
- [x] Add a `config_dir` option to galaxy.ini
- [x] Move `shed_tool_conf.xml` to its own config variable
- [x] Do not require `shed_*` configs to exist (they can be created as necessary)
- [x] Figure out what to do with all the mutable stuff (compiled templates, sqlite db, etc.)
- [x] ~~Install custom pip~~ not needed
- [x] Install dependencies
- [x] Figure out what to do with `tools/`
- [x] Distribute `templates/` and make it findable by Galaxy
- [x] Distribute `static/` and make it findable by Galaxy
- [ ] Other directories: `external_services/`, `tool-data/`, ...
- [x] Clean up the root directory and move packages to the top level
- [ ] Create a `galaxy-set-metadata` command to replace the tempfile method currently used
- [ ] Fix tool migrations
- [ ] Other things I'm not thinking of

natefoo committed:
Make Galaxy installable the Pythonic way via setup.py. Incomplete, but
if you install it and then run `galaxy` from the root of a Galaxy clone,
after installing all the dependencies (see the end of
./scripts/common_startup.sh), it should start.

mr-c commented:
:100: !!

natefoo committed:
Add some missing package data (although whether Galaxy can find it is
another story), fix classifiers.

martenson commented:
api tests on Jenkins report

```
Traceback (most recent call last):
  File "./scripts/manage_db.py", line 11, in <module>
    from migrate.versioning.shell import main
ImportError: No module named migrate.versioning.shell
```

martenson commented:
this needs to be rebased or dev merged into it before jenkins can test

natefoo commented:
@martenson Waiting for the tests to pass on #428 first.

natefoo committed:
Initial work toward separate galaxy-lib, galaxy-app, galaxy, and
tool-shed packages (see #1396). Lots of path fixing to be done still.

jmchilton commented on setup.py:
```diff
@@ -11,45 +12,123 @@
+if GALAXY_PACKAGE == "galaxy-lib":
+    requirements = []
+    base_packages = [
+        'galaxy.datatypes',
```
Datatypes depends on models so it shouldn't live in galaxy-lib in my opinion.

jmchilton commented:
What made you decide what to put in `galaxy-app` and what to put into `galaxy`? I feel like we should just figure out galaxy-lib first - and probably after this effort? After that I'd like to aim toward a design of:
- galaxy-lib (minimal required dependencies)
- galaxy-models (depends on sqlalchemy and galaxy-lib)
- galaxy-app (galaxy-models but no web stuff - someday it would be nice to run a webless galaxy process - e.g. job or workflow handlers using just this)
- galaxy-web (galaxy-app + controllers)

I would start from the bottom and introduce one or so a year as we refactor around these new distinctions? There are a think both practical and abstract design goals with each of these distinctions that I could go into in depth. None of this feels like something that should be rushed though.

natefoo committed:
Move tools, images, and templates in to lib/galaxy/ for inclusion in
wheels and packages.

natefoo committed:
Update client build for moved static directories

natefoo committed:
Improvements to distribution building

natefoo committed:
Fixes for paths expected to be relative to "galaxy root", which in the
"installed Galaxy" world is not the current working directory.

natefoo commented:
> What made you decide what to put in `galaxy-app` and what to put into `galaxy`?

"This seems like a web package", "This doesn't". Which is why it's almost certainly wrong.

So for now perhaps we should just have `galaxy-lib` and `galaxy`.

nsoranzo commented:
It would be nice to make the final tool migration (xref. #636) before this PR is merged.

natefoo committed:
Remove the galaxy-app package for now, and PEP 8 fixes

natefoo commented:
Per conversation with @jmchilton we are going to add a `config_dir` option to galaxy.ini so that we can stop assuming the location of everything is relative to the current working directory. Is there a sensible default for this? I suggest the directory containing galaxy.ini.

natefoo committed:
Remove the namespacing of galaxy packages for now, everything will just
live in the `galaxy` package. Install deps at install time.

natefoo committed:
Move some things from scripts/ to lib/ and some sample configs to
lib/galaxy/config/sample/

natefoo committed:
Add the config_dir and data_dir config options. Handle the following
files for "installed" galaxy:

- config/auth_conf.xml
- database/universe.sqlite
- database/files
- database/tool_dependency_dir
- database/shed_tools
- config/migrated_tools_conf.xml
- config/integrated_tool_panel.xml
- config/tool_sheds_conf.xml

In most cases these are read from internal defaults if not configured.
The exception (so far) is datatypes_conf.xml, which would be bad to
duplicate. It is instead read from
lib/galaxy/config/sample/datatypes_conf.xml.sample

This probably breaks tool migrations.

The deprecated database_file option has been removed.

natefoo committed:
Fix tests

natefoo committed:
Handle the following files for "installed" galaxy:

- config/data_manager_conf.xml
- config/external_service_types_conf.xml
- config/shed_data_manager_conf.xml
- config/shed_tool_data_table_conf.xml
- config/shed_data_manager_conf.xml
- config/tool_data_table_conf.xml
- config/workflow_schedulers_conf.xml
- tool-data (for shed-installed data tables)

shed_* are now created on demand when shed things are installed.

Installed Galaxy can now be started from outside the Galaxy root with
just a galaxy.ini. Some features will still be broken (static dbkeys,
display build sites). Some (dependency resolvers) are broken just
because I haven't gotten to fixing paths yet and will be easy to fix,
others (dbkeys) are more complicated because of complex paths and
relative-to-galaxy-root defaults (in datatypes_conf.xml).

natefoo committed:
Fixes for installed Galaxy

natefoo committed:
Fix for running from source

natefoo committed:
Fix galaxy-main

natefoo committed:
Fix linting errors

natefoo committed:
Handle remaining config and path moves for installable Galaxy

natefoo committed:
Add configurable options to allow for easy addition of static content
served via Galaxy, because in an installed context, the static directory
is not easily modifiable.

natefoo committed:
Ignore things generated as part of the wheel/client build process

natefoo committed:
Attempt to fix functional tests

natefoo committed:
Add a galaxy-config command to create galaxy.ini in the installed
context.

natefoo committed:
Provide path to additional samples in galaxy-config output

natefoo committed:
Fixes for tool migration when data_dir != ./data

natefoo committed:
Fix migrations and add the galaxy-manage-db command for installed
Galaxy. Could use a bit of help explaining how to use the
Galaxy-specific options that aren't passed on to sqlalchemy-migrate.

natefoo committed:
Don't duplicate requirements (setup.py reads requirements from
requirements.txt)

natefoo committed:
Do not attempt to create galaxy's root directory

natefoo committed:
Handle static paths in uWSGI in a very kludgey way (this should be
improved)

natefoo committed:
Split sample job configs back into two separate files for tests

natefoo committed:
Fix tox tests and api/framework tests should at least be able to run
now.

natefoo committed:
Add a --db-conn option to galaxy-config (Ideally in the future this
would be managed more automatically rather than having to specify the
whole URI) and ensure tool migrations are called with the correct
interpreter.

natefoo committed:
When running under uWSGI, sys.executable is the path to uwsgi. Attempt
to locate python.

natefoo committed:
(Slightly) refactor galaxy.config:Configuration for installable galaxy
changes so they do not have to be copy/pasted to the tool shed's
Configuration class.

natefoo committed:
Fix a rebase mistake

natefoo committed:
Fix paths to tool migrations

natefoo committed:
Do not use -c with manage_db.sh in Travis' startup tests, Galaxy can
find the sample config on its own.

natefoo committed:
Revert "Move tools, images, and templates in to lib/galaxy/ for inclusion in"

This reverts commit 77c2d153619a182d5be165fccd2cb85ab60b9725.

I'll move whatever needs to be moved after merging.

jmchilton commented:
I've never sat down and gone through this - but now that I have I stand by what I said the other day and even more so - this needing to broken into pieces 😅. This is too sprawling to ever get in as is IMO and has diverged really hard - and so many of the little pieces could have been great to have in for years.

natefoo committed:
Merge remote-tracking branch 'origin/dev' into installable

natefoo committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into installable

natefoo committed:
Fix Python 2.7 lint errors

natefoo committed:
Fix requirements symlink clobbered in merge

natefoo committed:
Config file fixes for installable Galaxy

natefoo committed:
Fix galaxy-manage-db

natefoo committed:
Restore tools/templates/static moves as symlinks

natefoo committed:
Fix static map paths

natefoo committed:
Don't require a config file

natefoo committed:
Fix datatypes_conf.xml.sample path for upload integration test

natefoo committed:
Move the galaxy metapackage to packages/ (incomplete)

natefoo committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into installable

natefoo committed:
Make default tool dependency dir relative to data dir

natefoo committed:
Fix lint errors

natefoo committed:
Excluded bundled tools from unit tests

natefoo committed:
Unit test fixes

natefoo committed:
Fix galaxy-main running from source now that setup,py is no longer in the root

natefoo committed:
Move remaining sample configs to lib/galaxy/config/sample

natefoo committed:
Symlink sample configs back to config/

jmchilton commented on Makefile:
```diff
@@ -171,6 +171,14 @@ client-eslint: node-deps  ## Run client linting
+# not rebuild client if unnecessary. But it's really only a problem in dev.
+# When this stuff goes into production it won't be as big of a deal.
+wheel-galaxy: client
+	python setup.py bdist_wheel
```
This file is gone right?

jmchilton commented on lib/galaxy/util/__init__.py:
```diff
@@ -1675,6 +1685,29 @@ def download_to_file(url, dest_file_path, timeout=30, chunk_size=2 ** 20):
+    return exe
+
+
+executable = _executable()
```
I'd prefer you move this into a new file. But maybe track that for after we merge this.

natefoo replied:
Where would you prefer?

jmchilton commented on lib/tool_shed/galaxy_install/migrate/check.py:
```diff
@@ -55,7 +56,9 @@ def verify_tools(app, url, galaxy_config_file=None, engine_options={}):
             if tool_shed_accessible:
                 # Automatically update the value of the migrate_tools.version database table column.
-                cmd = ['sh', 'manage_db.sh', 'upgrade', 'tools']
+                # FIXME: this doesn't seem right.
```
Note: this will be deleted 6 years from now and never fixed 😆

jmchilton commented on packages/meta/galaxy/project_galaxy_util.py:
```diff
@@ -0,0 +1,13 @@
+# -*- coding: utf-8 -*-
+
```
File name needs to be changed I think.

jmchilton commented on packages/meta/galaxy/project_galaxy_util.py:
```diff
@@ -0,0 +1,13 @@
+PROJECT_URL = "https://github.com/galaxyproject/galaxy"
+PROJECT_AUTHOR = 'Galaxy Project and Community'
+PROJECT_DESCRIPTION = 'Galaxy Server Metapackage'
+PROJECT_EMAIL = 'jmchilton@gmail.com'
```
We switched this to ``galaxy-committers@lists.galaxyproject.org``

natefoo replied:
And here I missed the chance to harass you about claiming authorship of all of Galaxy's code

jmchilton commented on packages/meta/setup.py:
```diff
@@ -0,0 +1,133 @@
+        galaxy-paster=galaxy.util.pastescript.serve:run
+        galaxy-main=galaxy.main:main
+        galaxy-config=galaxy.config.script:main
+        galaxy-manage-db=galaxy.model.orm.scripts:manage_db
```
This belong in galaxy-app after https://github.com/galaxyproject/galaxy/pull/8293 I think.

natefoo replied:
Added to the TODO

natefoo committed:
Fixes from PR review

natefoo committed:
Rework config locating logic (it was reading the sample again).

natefoo committed:
Stop trying to guess a web port from the config since most configs are not INI anymore anyway.

natefoo committed:
Merge branch 'dev' of https://github.com/galaxyproject/galaxy into installable

nsoranzo commented:
Wooot!

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -139,40 +97,112 @@ def resolve_path(path, root):
+            except Exception:
+                # Not an INI file
+                pass
+        self.config_dir = config_kwargs.get('config_dir', os.path.dirname(self.config_file or os.getcwd()))
```
@natefoo `os.path.dirname(os.getcwd())` is the parent of the current directory, so I think this may be wrong if `self.config_file` is None.
Also this is set to a different value on line 134 if `running_from_source` is True.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -860,10 +927,10 @@ def check(self):
             tool_configs.append(self.migrated_tools_config)
         for path in tool_configs:
-            if not os.path.exists(path):
+            if not os.path.exists(path) and path != self.shed_tool_conf:
```
@natefoo While trying to fix CI testing for BioBlend I noticed that:
- `path` may be a relative, e.g. `./config/shed_tool_conf.xml`, because it comes from `resolve_path(path, self.root)` in `_parse_config_file_options()` , where `self.root` is `.`
- `self.shed_tool_conf` is an absolute path, e.g. `/srv/galaxy/config/shed_tool_conf.xml`, because it comes from `resolve_path(kwargs.get('shed_tool_conf', 'shed_tool_conf.xml'), self.mutable_config_dir)` where `self.mutable_config_dir` is an absolute path.

Making `self.root` absolute seems a good way to solve this:
```diff
diff --git a/lib/galaxy/config/__init__.py b/lib/galaxy/config/__init__.py
index 92ab49fccd..aa1edd1f5f 100644
--- a/lib/galaxy/config/__init__.py
+++ b/lib/galaxy/config/__init__.py
@@ -100,7 +100,7 @@ def resolve_path(path, root):
 
 
 def find_root(kwargs):
-    root = kwargs.get('root_dir', '.')
+    root = os.path.abspath(kwargs.get('root_dir', '.'))
     return root
 
 
```

What do you think?

nsoranzo commented on lib/galaxy/tools/data_manager/manager.py:
```diff
@@ -37,11 +37,19 @@ def __init__(self, app, xml_filename=None):
+            try:
+                self.load_from_xml(self.app.config.shed_data_manager_config_file, store_tool_path=True)
+            except (OSError, IOError) as exc:
+                if exc.errno != errno.ENOENT or self.app.config.shed_data_manager_config_file_set:
```
@natefoo @ic4f I don't think `self.app.config.shed_data_manager_config_file_set` and `self.app.config.data_manager_config_file_set` are set anywhere.

natefoo replied:
IIRC this is dynamic, but I'll double check.

natefoo replied:
Yeah, they are set by `_parse_config_file_options()`

jdavcs replied:
I think they might get set here: 
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L162
Until this: https://github.com/galaxyproject/planemo/pull/950 was merged, that's where  `data_manager_config_file_set` was set (was passed with kwargs as empty string). 
That said, for future consideration, I think that block has too many special cases, which makes it not intuitive. Even a comment explaining the 'why' would help, I think.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -139,40 +97,112 @@ def resolve_path(path, root):
+        return os.path.join(self.sample_config_dir, path)
+
+    def _parse_config_file_options(self, defaults, listify_defaults, config_kwargs):
+        for var, defaults in defaults.items():
```
Probably it's not a bug, but reusing `defaults` variable name as loop variable is surely confusing.

jdavcs replied:
Yes!! I noticed it and hated it on first site.. I'll submit a quick PR (didn't want to trigger a code review over such a minor thing - but since you mention this.. so thanks :)

natefoo replied:
@ic4f thanks!

jdavcs replied:
@nsoranzo How about defaults >> vals? 
```
for var, vals in defaults.items():
    if config_kwargs.get(var, None) is not None:
        path = config_kwargs.get(var)
        setattr(self, var + '_set', True)
    else:
        for val in vals:
            if os.path.exists(resolve_path(val, self.root)):
                path = val 
                break
        else:
            path = vals[-1]
        setattr(self, var + '_set', False)
    setattr(self, var, resolve_path(path, self.root))
```

nsoranzo replied:
Maybe `values`/`value` to make it a bit more different from `var`.

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -139,40 +97,112 @@ def resolve_path(path, root):
+                setattr(self, var + '_set', False)
+            setattr(self, var, resolve_path(path, self.root))
+
+        for var, defaults in listify_defaults.items():
```
Here too.

martenson commented on lib/galaxy/tools/toolbox/base.py:
```diff
@@ -154,7 +160,21 @@ def _init_tools_from_config(self, config_filename):
+        try:
+            tool_conf_source = get_toolbox_parser(config_filename)
+        except (OSError, IOError) as exc:
+            for opt in ('shed_tool_conf', 'migrated_tools_config'):
```
fresh on dev this is called twice with `config_filename=/Users/marten/devel/git/temp/galaxy/config/migrated_tools_conf.xml` but once it has `getattr(self.app.config, opt)` equal to `/Users/marten/devel/git/temp/galaxy/config/shed_tool_conf.xml`.

so the `_dynamic_tool_confs` end up being 

```python
[{
'tool_path': '/Users/marten/devel/git/temp/galaxy/database/shed_tools',
 'create': '<?xml version="1.0"?>\n<toolbox tool_path="/Users/marten/devel/git/temp/galaxy/database/shed_tools">\n</toolbox>\n', 
'config_filename': '/Users/marten/devel/git/temp/galaxy/config/migrated_tools_conf.xml', 
'config_elems': []}]
```
 which seems wrong?

natefoo replied:
Can you create an issue and assign me?

martenson replied:
The issue is here: https://github.com/galaxyproject/galaxy/issues/8484 and you are assigned to it.

natefoo replied:
Ah, didn't realize this was the same thing, thanks.
